### ISSUE TITLE
Profiling points queries is *very* slow

### ISSUE THREAD

**Issue by nik9000** [elastic/elasticsearch#24440]
I've been trying to track down a performance problem and ran into a problem profiling points queries. They are just way, way slower to profile then they are to run. I'm not sure what is up but it certainly makes track down performance issues more difficult.

```
curl -XDELETE localhost:9200/test
curl -XPUT localhost:9200/test -d'{
  "settings": {
    "number_of_shards": 1,
    "number_of_replicas": 1
  }
}'

rm -f work
for i in $(seq 1 10000); do
  echo '{"index":{}}' >> work
  echo '{"number": 1, "term": "text"}' >> work
done

for i in $(seq 1 1000); do
  curl -s -XPOST localhost:9200/test/doc/_bulk --data-binary @work | cut -d , -f 1,2
done

curl -XPOST localhost:9200/_refresh

#slow
curl -XPOST 'localhost:9200/test/_search?size=1&pretty' -d'{
  "profile": true,
  "query": {
    "match": {
      "number": 1
    }
  }
}'

#fast
curl -XPOST 'localhost:9200/test/_search?size=1&pretty' -d'{
  "query": {
    "match": {
      "number": 1
    }
  }
}'

#fast
curl -XPOST 'localhost:9200/test/_search?size=1&pretty' -d'{
  "profile": true,
  "query": {
    "match": {
      "term": 1
    }
  }
}'

#fast
curl -XPOST 'localhost:9200/test/_search?size=1&pretty' -d'{
  "query": {
    "match": {
      "term": 1
    }
  }
}'
```

**Comment by polyfractal** [comment]
@jpountz do you think this is caused because the profiler uses [an alternate path](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/search/profile/query/ProfileWeight.java#L66) for bulk scoring?  Or is this perhaps something else?

**Comment by jpountz** [comment]
I would be surprised if that was the case since point queries are using the default bulk scorer. Only pure disjunctions, and maybe `match_all` should be affected by this alternate path for bulk scoring. I am maybe thinking of caching since the profile API disables the cache?

How much slower is slower in that case? Can you check it is not the case eg. with a `match_all` query? Could it be an issue with the fact that the benchmark does not include any warmup?

**Comment by jimczi** [comment]
I tested with `match_all` and it is effectively 10x slower with the profiler. Interestingly the point query you're running is transformed in a `match_all` query since all documents have the same value. This optimization is done on the PointRangeQuery so you're comparing a match all query with `profile` enabled or not. At first I though that this was because `ProfileWeight` does not use the new `scorerSupplier` path to build the inner scorer, that's an issue but for this specific example I think it's just that a `match_all` query is much faster with bulk scoring.

**Comment by jimczi** [comment]
Just to be clear `ProfileWeight` does not have to call the scorer supplier (it is called by the inner weight when building the scorer) so the main issue here is that a match all query is much slower when using the profiler.

**Comment by nik9000** [comment]
Does the rewriting time look large? I don't have the result sitting in front of me but I remember it was very high in the issue I was debugging.

**Comment by clintongormley** [comment]
Can we close this in favour of https://github.com/elastic/elasticsearch/pull/24565 ?

**Comment by nik9000** [comment]
> Can we close this in favour of #24565 ?

I don't think this is related to that one.

**Comment by clintongormley** [comment]
Won't sampling reduce the discrepancy between profile and unprofiled match-all execution?

**Comment by nik9000** [comment]
The one you linked isn't the sample issue. Also, I think this issue stands even with sampling because profiling changes the way that queries execute either with or without sampling.

**Comment by clintongormley** [comment]
Oh whoops - it's https://github.com/elastic/elasticsearch/issues/24799

> Also, I think this issue stands even with sampling because profiling changes the way that queries execute either with or without sampling.

Sure, measuring something changes its behaviour.  Question is, can we actually do anything about that?

**Comment by nik9000** [comment]
I mean, it *drastically* changes the behavior in this case. At least, that is my understanding. I wonder if we could warn users that the behavior has changed drastically if we detect that we're doing it. My understanding is that this particular change is so drastic as to invalidate the results.

**Comment by clintongormley** [comment]
> I mean, it drastically changes the behavior in this case.

Yes, and we can reduce how drastic it is with sampling. When I asked @jpountz about it, it didn't sound like there was much else we could do. I suppose one other possibility would be to just not profile match-all and assume that it is fast.

**Comment by jpountz** [comment]
Agreed that sampling might help with this issue. I think we should also work on eliminating the array access that we perform whenever we call `startTime` or `stopAndRecordTime`.

**Comment by clintongormley** [comment]
Discussed in FixItFriday.  There's not much more that we can do than to implement sampling, so I'm closing in favour of https://github.com/elastic/elasticsearch/issues/24799

**Comment by jpountz** [comment]
This issue should be addressed in the upcoming 5.6 and 6.0 releases via #25772.
